Loop/AppDelegate.swift

```diff
@@ -27,62 +27,108 @@ final class AppDelegate: UIResponder, UIApplicationDelegate, DeviceOrientationCo
     private var bluetoothStateManager: BluetoothStateManager!
 
     var window: UIWindow?
+    
+    var launchOptions: [UIApplication.LaunchOptionsKey: Any]?
 
     private var rootViewController: RootNavigationController! {
         return window?.rootViewController as? RootNavigationController
     }
-
-    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+    
+    private var isAfterFirstUnlock: Bool {
+        let fileManager = FileManager.default
+        do {
+            let documentDirectory = try fileManager.url(for: .documentDirectory, in: .userDomainMask, appropriateFor:nil, create:false)
+            let fileURL = documentDirectory.appendingPathComponent("protection.test")
+            guard fileManager.fileExists(atPath: fileURL.path) else {
+                let contents = Data("unimportant".utf8)
+                try? contents.write(to: fileURL, options: .completeFileProtectionUntilFirstUserAuthentication)
```

Comment:
IIRC, I think in my work around this I replaced this `try?` with a `try`.  Seems safe enough to do...WDYT?

Reply:
Yea, I don't think this works. This entire function is to see if we have access to the local file system (thus indicating after first unlock). 

Unless the `let documentDirectory = try fileManager.url` throws when it is locked (and I don't think it does), then the rest won't work. The `fileExists` call would return false (by definition, because we don't have access to the local file system before first unlock), the `guard` clause would be executed, the `try? contents.write` would do nothing (because of the `try?`) and it would return `true`. We'd need to change the `try?` to `try` to make this work.

Furthermore, I'd change `let contents = try? Data` to just `try` and let it throw. And then update the `log.error` message in the `catch` to be more accurate.

If `let documentDirectory = try fileManager.url` **does** throw when it is locked, then the rest of the function (and writing/reading the file) is unnecessary.

Reply:
@darinkrauss are you saying you don't think my suggestion works, or you think this whole code change won't work?

I'm not 100% sure I agree you can guarantee that `fileExists` will fail because the file system is unavailable.  See the following note in the [docs](https://developer.apple.com/documentation/foundation/filemanager/1410277-fileexists):

> It’s far better to attempt an operation (such as loading a file or creating a directory), check for errors, and handle those errors gracefully than it is to try to figure out ahead of time whether the operation will succeed.

Therefore, that's why I suggest `try` instead of `try?` when writing to the file.  Otherwise any potential error/exception thrown (admittedly very unlikely, but we never know across versions of iOS) will be ignored.  

(P.S. Way back when I was working on this, I am pretty sure I tested this code extensively and it worked...)

Reply:
Sorry I wasn't more clear. Yes, your fix is what will make it work. That is, changing the `try?` to `try`.

Reply:
The file existence check here does work before first unlock; this makes sense as individual files can have different access levels applied to them. The logic here is known to be working, and very hard to test; it requires specific minor versions of iOS that I don't know if we can actually install anymore. So I am loathe to make any changes.

@rickpasetto, you were not able to reproduce actually launching Loop before first unlock, iirc, so I don't know how you could've tested any changes to this.

The way it's written right now is that it will tend to launch if there are unexpected changes in iOS behavior. I think this is what we want. The preventative measure is in place and works if we see the same behavior that we saw in some versions of iOS 11 and early versions of 12. If we see other behavior, we'll have to deal with that. We have word from Apple that launching via bluetooth before unlock was a bug, and this code addresses that situation specifically.

Reply:
Huh, well, I don't see how it would logically, but perhaps it is throwing some non-Swift exception from `fileExists` or `write`, but then why would the Swift `catch` actually catch it? The difficulty I have is with `try? contents.write` in that it just silently fails. 🤷‍♂️

Reply:
@ps2 At some point, I was.  It was months ago now, though.  🤷‍♂️

Reply:
The `try? contents.write` isn't where we're checking for before/after first unlock; that's the code that is run if we're in the first run of the app, and we decide to create the file for later use. If it fails, we don't really care at that point; we can just continue launching the app.

Reply:
If that `try` fails, we can log an error if it fails (indeed, there's already a `catch` for this).  If we ever see it, we would know the behavior changed (e.g. an iOS version change modified the behavior somehow).